Fix missing sinks in attack graphs

jzelenjak opened the pull request:
## Description
### Issue 1
`traverse` method is missing some sink states. Namely, if there is a transition from a non-sink state to a sink state (which is defined for this non-sink node in the `final.json` file), then this sink state will be missed (i.e. drawn solid, see images below).

**Example 1:** states 427, 67 and 371 are sinks in `finalsinks.json` file generated by FlexFringe (this can also be seen in the fragments of S-PDFA, the count of these states is < `sink_count`=5), however in the AGs they are non-sinks (note: this example is also after implementing the fix with unknown in this PR: https://github.com/tudelft-cda-lab/SAGE/pull/12). Left AG is before the fix, right AG is after the fix.

![pr1 1](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/e5470e88-0144-4a98-bb7d-a2240cbdf41f)
![pr1 2](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/d6d65e29-4030-4898-b60c-c463912bdd43)
![pr1 3](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/cc0c6b67-6196-4b5e-8353-498b9dd6947e)
![pr1 4](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/ac56fab6-4bfe-4dd2-9fca-a2f0e97a1507)

**Example 2:** state 82 is a non-sink state(in the left image before the fix), while in `finalsinks.json` file generated by FlexFringe it is a sink

![pr1 5](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/58423098-061d-41f1-80e0-ffbcd9632d79)

**Example 3:** states 80 and 81 a non-sink states (in the left image before the fix), while in `finalsinks.json` file generated by FlexFringe they are sinks
![pr1 6](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/8adcf64a-173d-4b54-ac6d-04a90396a0b6)

**Example 4:** state 70 is a non-sink state (in the left image before the fix), while in `finalsinks.json` file generated by FlexFringe it is a sink

![pr1 7](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/0d73f26b-4815-49ba-ac23-572179899e60)

### Issue 2

If a sequence ends with a sink, then it will not be added to the `sev_sinks`, as `state` will not be an empty string. In `final.json` file, there are transitions from non-sinks to sinks, but sinks themselves are not defined (they are in `finalsinks.json` file). So, SAGE misses sinks that are at the end of a sequence.

![pr1 8](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/2674f17f-1b68-42b7-8d53-d6c658dd144f)
![pr1 9](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/9467787a-9bb5-426b-9d02-8ade34ba4ef2)

### Issue 3

Sometimes there is a transition from a red non-sink state to a blue sink state. However, in the default `spdfa-config.ini` `printblue` parameter is set to `0`. This results in a missing transition in `final.json` file, because of which `ID: -1` will be assigned for medium-severity states, even though they do have an ID. This is especially a problem when there are multiple such states in an AG, as they will be merged to one state with ID: -1. Furthermore, it hinders interpretability.

**Before (fragment):**

![pr1 10](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5d1d8138-6885-488d-9015-2dac8c1727a9)

**After (fragment):**
![pr1 11](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/4e25f305-e3b3-4b6f-b62a-ae0b9d53c406)

**Another example + the fragments of S-PDFA (note: here 287 and 289 are the same nodes, as well as 1451 and 1437 - the screenshot of the S-PDFA merge was taken after implementing all the fixes, including PR https://github.com/tudelft-cda-lab/SAGE/pull/12 and PR https://github.com/tudelft-cda-lab/SAGE/pull/11**

![pr1 12](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/34a9e707-6469-40ff-937c-1d4b3e7f4b77)
![pr1 13](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/d2592c09-f484-4667-a832-df43e00d611b)

### Issue 4

In `make_AG` method, in three places, it is checked whether states are sinks and if they are, they are made dotted. This is done using `endwith` method, namely: `obj.endswith(sink)`. However, this might result in a false positive if `sink` is a string substring of the last part of the objective variant. For example, the condition `obj.endswith(sink)` will evaluate to True if `sink=75` and `obj=”DATA_EXFILTRATION|http|375`, which will lead to state 375 coloured dashed, whereas it is not a sink.

![pr1 14](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/066697c0-8874-4a38-8228-3d540ab18092)
![pr1 15](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/c258196e-3732-48d8-a028-95cef5248a73)
![pr1 16](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/407f44d3-6dc7-4c42-815e-0fbc234ee1e7)

## Proposed fixes

First, replace `endswith` method with strict equals comparison after splitting an objective or a vertex name.

![pr1 17](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/2b2ef722-7d3a-41e7-ac51-b1d5ea8ff32b)

Second, at the end of `traverse` method, add a check to add a state in the `sev_sinks` set. This fix will prevent skipping sinks that are heads of edges with tails being non-sinks (defined in the main model). Note that if `ID: -1`, for low-severity sinks, they will not be added, as `-1` is never in `sinks` (the intended behaviour).

![pr1 18](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/fbffb9e3-1f6a-4c9f-8553-2176251d130f)

Third, set `printblue=1` by default to `spdfa-config.json` file. This will remove all the `ID: -1` from medium-severity states and put the actual IDs as in the files generated by FF.

## Result

1. Resulting AGs are the same when it comes to node names (excluding state IDs) and edges, except for AG for `10.0.0.1|NETWORKDOS|mswbtserver`, where before there was one `ACCOUNT MANIPULATION|-1` node and after there are two such nodes: `ACCOUNT MANIPULATION|snmp|ID: 371` and `ACCOUNT MANIPULATION|snmp|ID:863`.

![pr1 19](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/bc8aba35-841d-4c3f-a9d2-96f8fe4c9291)

2. Plenty of lost sinks have been found (note: these stats are before the fixes in PRs https://github.com/tudelft-cda-lab/SAGE/pull/11 and https://github.com/tudelft-cda-lab/SAGE/pull/12)

![pr1 20](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/4d358634-1e62-43be-a523-bf97766d3de4)

3. All medium-severity states with `ID: -1` have disappeared:

```bash
[jegor@arch SAGE-fork]$ find before-2017AGs/ -type f -name '*.dot' | xargs gvpr 'N { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | grep -- 'ID: -1' | wc -l
71
[jegor@arch SAGE-fork]$ find before-2017AGs/ -type f -name '*.dot' | xargs gvpr 'N { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | grep -- 'ID: -1' | sort -u | wc -l
10
[jegor@arch SAGE-fork]$ find after-2017AGs/ -type f -name '*.dot' | xargs gvpr 'N { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | grep -- 'ID: -1' | wc -l
0
[jegor@arch SAGE-fork]$ find before-2018AGs/ -type f -name '*.dot' | xargs gvpr 'N { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | grep -- 'ID: -1' | wc -l
2
[jegor@arch SAGE-fork]$ find after-2018AGs/ -type f -name '*.dot' | xargs gvpr 'N { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | grep -- 'ID: -1' | wc -l
0
```

4. With the following pipelines, we have verified all our findings. First, we get all unique sinks in CPTC-2017 (analogously for CPTC-2018) for AGs before and after the fixes. Then we take the ones that are present only after the fix (`comm -13 …`), and extract their IDs (`sed …`). At the same time, using `jq` we take the `before-2017.txt.ff.finalsinks.json` file (which is the same as `after-2017.txt.ff.finalsinks.json` file) and print IDs of only those nodes that are sinks. Finally, we take the intersection between the previously found IDs (`comm -12 …`) and count them (` … | wc -l`). The result is equal to the number of such sink IDs that are only present after the fix (i.e. lost by the original version of SAGE before the fix).

```bash
# All new-found sinks in 2017 are indeed sinks
[jegor@arch SAGE-fork]$ sinks_before_2017=$(find before-2017AGs/ -type f -name '*.dot' | xargs grep -F -l "dotted" | xargs gvpr 'N [ $.style == "dotted" || $.style == "filled,dotted" || $.style == "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i)
[jegor@arch SAGE-fork]$ sinks_after_2017=$(find after-2017AGs/ -type f -name '*.dot' | xargs grep -F -l "dotted" | xargs gvpr 'N [ $.style == "dotted" || $.style == "filled,dotted" || $.style == "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i)
[jegor@arch SAGE-fork]$ echo -e "$sinks_before_2017" | wc -l
66
[jegor@arch SAGE-fork]$ echo -e "$sinks_after_2017" | wc -l
141
[jegor@arch SAGE-fork]$ new_sinks_2017=$(comm -13 <(echo -e "$sinks_before_2017") <(echo -e "$sinks_after_2017"))
[jegor@arch SAGE-fork]$ echo -e "$new_sinks_2017" | wc -l
75
[jegor@arch SAGE-fork]$ diff -q before-2017.txt.ff.finalsinks.json after-2017.txt.ff.finalsinks.json 
[jegor@arch SAGE-fork]$ all_sinks_2017=$(jq '.nodes[] | select(.issink==1) | .id' before-2017.txt.ff.finalsinks.json | sort)
[jegor@arch SAGE-fork]$ comm -12 <(echo -e "$new_sinks_2017" | sed 's/^.*ID: \([0-9-]\+\)$/\1/' | sort) <(echo -e "$all_sinks_2017") | wc -l
75

# All new-found sinks in 2018 are indeed sinks
[jegor@arch SAGE-fork]$ sinks_before_2018=$(find before-2018AGs/ -type f -name '*.dot' | xargs grep -F -l "dotted" | xargs gvpr 'N [ $.style == "dotted" || $.style == "filled,dotted" || $.style == "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i)
[jegor@arch SAGE-fork]$ sinks_after_2018=$(find after-2018AGs/ -type f -name '*.dot' | xargs grep -F -l "dotted" | xargs gvpr 'N [ $.style == "dotted" || $.style == "filled,dotted" || $.style == "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i)
[jegor@arch SAGE-fork]$ echo -e "$sinks_before_2018" | wc -l
58
[jegor@arch SAGE-fork]$ echo -e "$sinks_after_2018" | wc -l
104
[jegor@arch SAGE-fork]$ new_sinks_2018=$(comm -13 <(echo -e "$sinks_before_2018") <(echo -e "$sinks_after_2018"))
[jegor@arch SAGE-fork]$ echo -e "$new_sinks_2018" | wc -l
46
[jegor@arch SAGE-fork]$ diff -q before-2018.txt.ff.finalsinks.json after-2018.txt.ff.finalsinks.json
[jegor@arch SAGE-fork]$ all_sinks_2018=$(jq '.nodes[] | select(.issink==1) | .id' before-2018.txt.ff.finalsinks.json | sort)
[jegor@arch SAGE-fork]$ comm -12 <(echo -e "$new_sinks_2018" | sed 's/^.*ID: \([0-9-]\+\)$/\1/' | sort) <(echo -e "$all_sinks_2018") | wc -l
46
```

Furthermore,  in a similar way as above, we have verified all sinks found after all fixes are indeed sinks in FlexFringe and all non-sinks with defined IDs (i.e. ID has not been removed) are indeed non-sinks in FlexFringe:

```bash
# All found sinks in 2017 are indeed sinks (after all fixes)
[jegor@arch SAGE-fork]$ sinks_after_2017=$(find after-2017AGs/ -type f -name '*.dot' | xargs grep -F -l "dotted" | xargs gvpr 'N [ $.style == "dotted" || $.style == "filled,dotted" || $.style == "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i)
[jegor@arch SAGE-fork]$ all_sinks_2017=$(jq '.nodes[] | select(.issink==1) | .id' before-2017.txt.ff.finalsinks.json | sort)
[jegor@arch SAGE-fork]$ echo -e "$sinks_after_2017" | wc -l
141
[jegor@arch SAGE-fork]$ comm -12 <(echo -e "$sinks_after_2017" | sed 's/^.*ID: \([0-9-]\+\)$/\1/' | sort) <(echo -e "$all_sinks_2017") | wc -l
141

# All found sinks in 2018 are indeed sinks (after all fixes)
[jegor@arch SAGE-fork]$ sinks_after_2018=$(find after-2018AGs/ -type f -name '*.dot' | xargs grep -F -l "dotted" | xargs gvpr 'N [ $.style == "dotted" || $.style == "filled,dotted" || $.style == "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i)
[jegor@arch SAGE-fork]$ all_sinks_2018=$(jq '.nodes[] | select(.issink==1) | .id' before-2018.txt.ff.finalsinks.json | sort)
[jegor@arch SAGE-fork]$ echo -e "$sinks_after_2018" | wc -l
104
[jegor@arch SAGE-fork]$ comm -12 <(echo -e "$sinks_after_2018" | sed 's/^.*ID: \([0-9-]\+\)$/\1/' | sort) <(echo -e "$all_sinks_2018") | wc -l
104
```

```bash
# All non-sinks with IDs in 2017 are indeed non-sinks (after all fixes)
[jegor@arch SAGE-fork]$ non_sinks_with_ids_after_2017=$(find after-2017AGs/ -type f -name '*.dot' | xargs gvpr 'N [ $.style != "dotted" && $.style != "filled,dotted" && $.style != "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i | grep 'ID: ')
[jegor@arch SAGE-fork]$ echo -e "$non_sinks_with_ids_after_2017" | wc -l
28
[jegor@arch SAGE-fork]$ all_non_sinks_after_2017=$(jq '.nodes[] | select(.issink==0) | .id' after-2017.txt.ff.final.json | sort -u)
[jegor@arch SAGE-fork]$ comm -12 <(echo -e "$non_sinks_with_ids_after_2017" | sed 's/^.*ID: \([0-9-]\+\)$/\1/' | sort -u) <(echo -e "$all_non_sinks_after_2017") | wc -l
28

# All non-sinks with IDs in 2018 are indeed non-sinks (after all fixes)
[jegor@arch SAGE-fork]$ non_sinks_with_ids_after_2018=$(find after-2018AGs/ -type f -name '*.dot' | xargs gvpr 'N [ $.style != "dotted" && $.style != "filled,dotted" && $.style != "dotted,filled" ] { print(gsub(gsub($.name, "\r"), "\n", " | ")); }' | sort -u | uniq -i | grep 'ID: ')
[jegor@arch SAGE-fork]$ echo -e "$non_sinks_with_ids_after_2018" | wc -l
16
[jegor@arch SAGE-fork]$ all_non_sinks_after_2018=$(jq '.nodes[] | select(.issink==0) | .id' after-2018.txt.ff.final.json | sort -u)
[jegor@arch SAGE-fork]$ comm -12 <(echo -e "$non_sinks_with_ids_after_2018" | sed 's/^.*ID: \([0-9-]\+\)$/\1/' | sort -u) <(echo -e "$all_non_sinks_after_2018") | wc -l
16
```

jzelenjak committed:
Fix missing sinks in traverse and make_ag methods

jzelenjak committed:
Add spdfa-config.ini

jzelenjak committed:
Add a check to exclude low-severity sinks from sev_sinks set

azqanadeem commented:
The .ini file is not necessary here, it is already a part of the docker container. Please remove unnecessary changes in this PR such that it works directly with the docker container as well.

jzelenjak committed:
Remove spdfa-config.ini for consistency with the docker container

jzelenjak committed:
Revert path_to_ini

jzelenjak commented:
I have removed the config file and changed back the `path_to_ini`. I thought that the `spdfa-config.ini` file was missing, since we had to take it from the `docker` branch.
